README.rst
relstorage_packer
=================

Packs a ZODB in a *history-free* PostgreSQL RelStorage with blobs in
filesystem.


Overview
--------

This script works also for very large Relstorage ZODBs with several million
objects. The original pack script took several days and consumed lots of RAM.
So there was need to accelerate the process of packing.

This script does not consume relevant amounts of RAM, runs much faster than the
original. Where the old took 3.5 days only for analysis it takes now about 6
hours. On subsequent runs it only processes changes after last run: it
considers only transactions newer than last processed transaction of the prior
run.

At time of writing processing 44mio objects takes initially about 3-6h
depending on hardware and configuration of Postgresql.

The script creates an inverse object graph, this takes little extra space in DB.


Limitations
-----------

At time of development the critical production environment was a postgresql
database running relstorage with blobs stored on a fileserver in history free
mode. So this is implemented.

I'am sure its easily possible to make this work on MySQL and Oracle too.
Also considering blobs inside DB is for sure possible.

I'am not sure if this way of cleanup makes sense for non-history-free mode. At
least it needs a lot of love and understanding of ZODB to refactor and
implement.

Contributions are welcome!


Usage
-----

Create a configuration file. It is the same as used in classical pack script
deployed with Relstorage::

    <relstorage>
        create-schema false
        keep-history false
        shared-blob-dir true
        blob-dir var/blobstorage
        commit-lock-timeout 600
        <postgresql>
            dsn dbname='test_site' host='127.0.0.1' user='zodb' password='secret'
        </postgresql>
    </relstorage>

After installation a script ``relstorage_pack`` is available::

    Usage: relstorage_pack config_file

    Fast ZODB Relstorage Packer for history free PostgreSQL
    
    Options:
      -h, --help     show this help message and exit
      -i, --init     Removes all reference counts and starts from scratch.
      -v, --verbose  More verbose output, includes debug messages.

When running first time with your database pass ``--init`` as parameter. This
drops and recreates the packing table.


How it works
============

At first run it creates a table ``object_inrefs`` used for inverse reference
counting. The table has:

``zoid BIGINT NOT NULL,``
    this is the object id where incoming references are counted  for

``tid BIGINT NOT NULL CHECK (tid > 0)``
    transaction id of the zoid

``inref BIGINT,``
    the object id of the incoming reference OR
    the same as zoid.

``numinrefs BIGINT NOT NULL DEFAULT 0,``
    if zoid==inref this is the counter field, otherwise it is not relevant.

So this table is used for two different things:

1) keeping track of the incoming references

2) counting the incoming references.

The code runs in three main phases:

initial preparation phase
    creates missing tables, cleans object_inrefs table and runs through all
    transactions in order to count and record all transactions.

subsequent runs preparation phase
    1) starts at last know tid and then runs through all new
       transactions in order to count and record changes of new transactions.
    2) for each new transaction zoid (current) check also if there where
       references gone meanwhile. so get all prior filed references of current
       and remove any not valid anymore. For each removed decrement the counter
       on ``object_inrefs`` where zoid=reference and inref=reference.

cleanup phase
    1) select an orphan, a zoid with no incoming refs
    2) get all zoids referenced by this orphan
    3) for each of this reference delete the entry from ``object_inrefs`` where
       inref=orphan and zoid=reference
    4) decrement counter on the entry where zoid=reference and inref=reference
    5) delete the entry with the orphaned zoid from ``object_state`` (real data)
    6) start with (1) unless theres no orphan any more.


Source Code
===========

The sources are in a GIT DVCS with its main branches at
`github <http://github.com/bluedynamics/relstorage_packer>`_.

We'd be happy to see many forks and pull-requests to make this package even
better.

Using integrated buildout and testing
-------------------------------------

Testing this code i not easy and writing good tests is a task to be done.
At the moment you can try the code bu running a
postgres database on localhost (unless you want to change ``buildout.cfg``).
Then run as database-user (named ``postgres`` on debian) the commands::

    psql -c "CREATE USER zope WITH PASSWORD 'secret';"
    psql -c "CREATE DATABASE relstorage_packer_test OWNER zope;"
    psql -c "REVOKE connect ON DATABASE relstorage_packer_test FROM PUBLIC;"
    psql -c "GRANT connect ON DATABASE relstorage_packer_test TO zope;"
 
Next (because of my laziness) run ``./bin/instance start`` which starts a Plone.
Add a Plone Site, add and delete some content to fill the database with
something to pack.

Next run the packer.

If you dont like this: pull requests are always welcome.

Contributors
============

- Jens W. Klein <jens@bluedynamics.com> (Maintainer)

Thanks to Robert Penz for some good ideas at our Linux User Group Tirol Meeting.
Also thanks to Shane Hathaway for ``Relstorage`` and Jim Fulton for ZODB and
``zc.zodbdgc`` (which unfortunately does not work with Relstorage).


HISTORY.rst

History
=======

2.1 (2014-02-19)
----------------

- refactored logging, because it was always in my way when changing other parts 
  of the code.
  [jensens, 2014-02-19]

- after long running connections postgresql takes lots of RAM. So we reconnect
  every 5000 cycles (TID analyzing or ZOID removal).
  [jensens, 2014-02-18]

- we had a whole bunch of ``idle in transaction (aborted)`` postgres
  processes running after packing. This resulted in in an ``OperationalError:
  out of shared memory HINT: You might need to increase
  max_pred_locks_per_transaction.`` Error. As a result I refactored the
  transaction handling and rollback and use explicit commit instead of using
  the relstorage ``storage._with_store``. Now this part is very controlled
  and not the source of hanging connections w/o rollback.
  [jensens, 2014-02-18]


2.0.2 (2014-02-05)
------------------

- also support storages w/o blobstorage
  [jensens, 2014-02-05]


2.0.1 (2014-02-03)
------------------

- unlock in a finally to really unlock an failure
  [jensens, 2014-02-03]

- use non-zero exit code if lock could not be aquired.
  [saily]


2.0
---

- refactored the way of collecting and using the reference counts. faster now.
  [jensens, 2014-01-11]

1.0
---

- started package
  [jensens, 2013-11-23]


src/vdexcsv/tests.py
import unittest
import doctest 
from pprint import pprint
from interlude import interact
import lxml.etree as etree

optionflags = doctest.NORMALIZE_WHITESPACE | \
              doctest.ELLIPSIS | \
              doctest.REPORT_ONLY_FIRST_FAILURE

TESTFILES = [
    'api.txt',
    'script.txt',
]

def fxml(xml):
    et = etree.fromstring(xml)
    return etree.tostring(et, pretty_print=True)

def pxml(xml):
    print fxml(xml)

def test_suite():
    return unittest.TestSuite([
        doctest.DocFileSuite(
            file, 
            optionflags=optionflags,
            globs={'interact': interact,
                   'pprint': pprint,
                   'pxml': pxml},
        ) for file in TESTFILES
    ])

if __name__ == '__main__':                                   #pragma NO COVERAGE
    unittest.main(defaultTest='test_suite')                  #pragma NO COVERAGE

src/relstorage_packer/tests.py
import unittest
import doctest
from pprint import pprint
from interlude import interact
import lxml.etree as etree

optionflags = doctest.NORMALIZE_WHITESPACE | \
              doctest.ELLIPSIS | \
              doctest.REPORT_ONLY_FIRST_FAILURE

TESTFILES = [
    'api.txt',
    'script.txt',
]


def fxml(xml):
    et = etree.fromstring(xml)
    return etree.tostring(et, pretty_print=True)


def pxml(xml):
    print fxml(xml)


def test_suite():
    return unittest.TestSuite([
        doctest.DocFileSuite(
            file_,
            optionflags=optionflags,
            globs={'interact': interact,
                   'pprint': pprint,
                   'pxml': pxml},
        ) for file_ in TESTFILES
    ])


if __name__ == '__main__':  # pragma NO COVERAGE
    unittest.main(defaultTest='test_suite')  # pragma NO COVERAGE


src/vdexcsv/api.py
import csv
from lxml import etree
from collections import OrderedDict

NSVDEX = 'http://www.imsglobal.org/xsd/imsvdex_v1p0'
NSXSI  = "http://www.w3.org/2001/XMLSchema-instance"
NSMAP = {
    None: NSVDEX,
    'xsi': NSXSI
}

def vtag(tag, ns=NSVDEX):
    return "{%s}%s" % (ns, tag) 

class CSV2VDEX(object):
    
    def __init__(self, vid, names, infile, outfile, 
                 startrow=0, colkey=0, colstartvalue=1, langs='en',
                 dialect='excel', delimiter=';', treevocabdelimiter='.',
                 ordered=True, encoding='utf-8'):
        self.vid = vid
        if isinstance(langs, basestring):
            langs = [_.strip() for _ in langs.split(',')]
        if isinstance(names, basestring):
            names = [_.strip() for _ in names.split(',')]
        self.names = OrderedDict(zip(langs, names))
        self.infile = infile
        self.outfile = outfile
        self.startrow = startrow
        self.colkey = colkey
        self.colstartvalue = colstartvalue
        self.langs = langs
        self.delimiter = delimiter
        self.treevocabdelimiter = treevocabdelimiter
        self.ordered = ordered
        if dialect not in csv.list_dialects():
            raise ValueError, "given csv dialect '%s' is unknown. " % dialect +\
                              "pick one of theses: %s" % csv.list_dialects() 
        self.dialect = dialect
        self.encoding = encoding
        
    @property
    def _fields(self):
        maxlen = max(self.colstartvalue+len(self.langs), self.colkey+1)
        fields = ['__genkey-%s' % _ for _ in range(0, maxlen)]
        fields[self.colkey] = 'key'
        for idx in range(0, len(self.langs)): 
            if self.colstartvalue+idx == self.colkey:
                raise ValueError, 'key column is in same range as value columns'
            fields[self.colstartvalue+idx] = self.langs[idx]
        return fields 
        
    @property
    def _csvdict(self):
        infile = self.infile
        if isinstance(infile, basestring):
            infile = open(infile, 'Ur')
        value = csv.DictReader(infile, fieldnames=self._fields,
                               dialect=self.dialect, 
                               delimiter=self.delimiter)
        tree = OrderedDict()
        rowcount = -1
        for item in value:
            rowcount += 1
            if rowcount < self.startrow:
                continue
            parts = item['key'].split(self.treevocabdelimiter)
            length = len(parts)
            branch = tree
            for idx in range(0, length):
                part = parts[idx]
                if idx < length-1:
                    branch = branch[part][0]
                    continue
                values = [item[_].decode(self.encoding) for _ in self.langs]
                branch[part] = OrderedDict(), values, item['key']
        return tree
    
    @property
    def _xml(self):
        root = etree.Element(vtag("vdex"), nsmap=NSMAP)
        root.attrib[vtag("schemaLocation", ns=NSXSI)] = \
            "http://www.imsglobal.org/imsvdex_v1p0 imsvdex_v1p0.xsd"
        root.attrib['profileType'] = 'lax'
        root.attrib['orderSignificant'] = str(self.ordered).lower()
        vid = etree.SubElement(root, vtag('vocabIdentifier'))
        vid.text = self.vid
        cap = etree.SubElement(root, vtag('vocabName'))
        for lang in self.names:
            langstring = etree.SubElement(cap, vtag('langstring'))
            langstring.text = self.names[lang]
            langstring.attrib['language'] = lang                     
        def treeworker(tree, parent):
            for key in tree:
                subtree, values, longkey = tree[key]
                term = etree.SubElement(parent, vtag('term'))
                termid = etree.SubElement(term, vtag('termIdentifier'))
                termid.text = longkey
                caption = etree.SubElement(term, vtag('caption'))
                for idx in range(0, len(values)):
                    langstring = etree.SubElement(caption, vtag('langstring'))
                    langstring.text = values[idx]
                    langstring.attrib['language'] = self.langs[idx]                     
                treeworker(subtree, term)
        treeworker(self._csvdict, root)            
        return etree.tostring(root, pretty_print=True)                        

    def __call__(self):
        with open(self.outfile, 'w') as outfile:
            outfile.write(self._xml)

src/vdexcsv/script.py
import argparse
from vdexcsv import api 

parser = argparse.ArgumentParser(description='Converts CSV files to VDEX XML')
parser.add_argument('--languages', '-l', nargs='?', default='en', 
                    help='Comma separated list of ISO-language codes. ' 
                    'Default: en')
parser.add_argument('--startrow', '-r', nargs='?', type=int, default=0, 
                    help='number of row in CSV file where to begin reading, '
                        'starts with 0, default 0.')
parser.add_argument('--keycolumn', '-k', nargs='?', type=int, default=0, 
                    help='number of column with the keys of the vocabulary, '
                         'start with 0, default 0.')
parser.add_argument('--startcolumn', '-s', nargs='?', type=int, default=1, 
                    help='number of column with the first langstring of the ' 
                         'vocabulary. It assume n + number languages of columns'
                         ' after this, starts counting with 0, default 1.')
parser.add_argument('--ordered', '-o', nargs='?', type=bool, default=True, 
                    help='Wether vocabulary is ordered or not, Default: True')
parser.add_argument('--dialect', nargs='?', default='excel', 
                    help='CSV dialect, default excel.')
parser.add_argument('--csvdelimiter', nargs='?', default=';', 
                    help='CSV delimiter of the source file, default colon.')
parser.add_argument('--treedelimiter', nargs='?', default='.', 
                    help='Delimiter used to split the key the vocabulary into '
                         'a path to determine the position in the tree, '
                         'default dot.')
parser.add_argument('--encoding', '-e', nargs='?', default='utf-8', 
                    help='Encoding of input file. Default: utf-8')
parser.add_argument('id', nargs=1,
                   help='unique identifier of vocabulary')
parser.add_argument('name', nargs=1, 
                   help='Human readable name of vocabulary. If more than one '
                        'language is given separate each langstring by a colon '
                        'and provide same order as argument --languages')
parser.add_argument('source', nargs=1,
                   help='CSV file to read from')
parser.add_argument('target', nargs=1,
                   help='XML target file')

def run():
    ns = parser.parse_args()
    api.CSV2VDEX(ns.id[0], ns.name[0], ns.source[0], ns.target[0], 
                 ns.startrow, ns.keycolumn, ns.startcolumn, ns.languages, 
                 ns.dialect, ns.csvdelimiter, ns.treedelimiter, 
                 ns.ordered, ns.encoding)()

src/relstorage_packer/utils.py
import logging
from StringIO import StringIO
import ZConfig
from ZODB.serialize import referencesf
from ZODB.utils import u64

log = logging.getLogger("utils")

schema_xml = """
<schema>
  <import package="ZODB"/>
  <import package="relstorage"/>
  <multisection type="ZODB.storage" attribute="storages" />
</schema>
"""

def dbcommit(func):
    """decorator with save commit

    closes cursor!
    """

    def _wrapper(connection, cursor, *args, **kw):
        try:
            result = func(cursor, *args, **kw)
            cursor.close()
            connection.commit()
        except:
            connection.rollback()
            raise
        return result
    return _wrapper


def get_storage(config_file):
    schema = ZConfig.loadSchemaFile(StringIO(schema_xml))
    config, dummy = ZConfig.loadConfig(schema, config_file)
    if len(config.storages) < 1:
        raise ValueError('No storages configured')
    connection = config.storages[0]
    if connection.config.keep_history:
        raise RuntimeError('Packing does not support history keeping storages')
    name = '%s (%s)' % ((connection.name or 'storage'),
                        connection.__class__.__name__)
    log.info("Opening %s...", name)
    storage = connection.open()
    log.info("Successfully openend %s", storage.getName())
    if 'PostgreSQLAdapter' not in storage.getName():
        raise RuntimeError('Only PostgreSQL databases are supported')
    return storage

def get_conn_and_cursor(storage):
    adapter = storage._adapter
    return adapter.connmanager.open()

def get_cursor(connection):
    return connection.cursor()

def get_references(state):
    """Return the set of OIDs the given state refers to."""
    refs = set()
    if state:
        for oid in referencesf(str(state)):
            refs.add(u64(oid))
    return refs


src/relstorage_packer/refcount.py
"""relstorage_packer - reference numinrefs process"""
from .utils import dbcommit
from .utils import get_conn_and_cursor
from .utils import get_references
from .utils import get_storage
from ZODB.utils import p64
from psycopg2 import InterfaceError
import datetime
import logging
import optparse
import os
import shutil
import sys
import time

WAIT_DELAY = 1
CYCLES_TO_RECONNECT = 5000
LOG_INTERVAL_SECS = 1

log = logging.getLogger("pack")
log.setLevel(logging.INFO)

################################################################################
# Locking

@dbcommit
def aquire_lock(cursor):
    """
    try to acquire a relstorage_packer lock, if not possible log error and exit
    with 1
    """
    log.info("Acquiring relstorage_packer lock")
    cursor.execute("SELECT pg_try_advisory_lock(23)")
    locked = cursor.fetchone()[0]
    if not locked:
        log.error("Impossible to get relstorage_packer Lock. Exit.")
        exit(1)

@dbcommit
def release_lock(cursor):
    """release relstorage_packer lock
    """
    log.info("Releasing relstorage_packer lock")
    cursor.execute("SELECT pg_advisory_unlock(23)")


################################################################################
# Initialization

@dbcommit
def init_table(cursor):
    log.info("Create table object_inrefs (drop existing).")
    stmt = """
    DROP TABLE IF EXISTS object_inrefs;
    CREATE TABLE object_inrefs (
        zoid       BIGINT NOT NULL,
        tid        BIGINT NOT NULL CHECK (tid > 0),
        inref      BIGINT,
        numinrefs    BIGINT NOT NULL DEFAULT 0,
        PRIMARY KEY(zoid, inref)
    );
    CREATE INDEX object_inrefs_tid  ON object_inrefs (tid);
    CREATE INDEX object_inrefs_refs ON object_inrefs (inref);
    CREATE INDEX object_inrefs_numinrefs ON object_inrefs (numinrefs);
    """
    cursor.execute(stmt)

    stmt = """
    CREATE OR REPLACE
        FUNCTION add_inref(
            vfrom BIGINT, vto BIGINT, vtid BIGINT
        )
    RETURNS void
    AS $$
    DECLARE
        existing_ref boolean;
    BEGIN
        SELECT true
            INTO existing_ref
                FROM object_inrefs
                WHERE zoid = vto AND inref = vfrom;
        IF existing_ref THEN
            UPDATE object_inrefs
                SET tid = vtid
                WHERE zoid = vto AND inref = vto;
        ELSE
            INSERT INTO object_inrefs VALUES(vto, vtid, vfrom, 0);
            UPDATE object_inrefs
                SET numinrefs = numinrefs + 1
                WHERE zoid = vto AND inref = vto;
        END IF;
        RETURN;
    END;
    $$ LANGUAGE plpgsql;
    """
    cursor.execute(stmt)
    _add_ref(cursor, 0, 0, 1)
    _add_ref(cursor, -1, 0, 1)


################################################################################
# Fetching of Transaction Ids to be processed

def tid_boundary(cursor):
    """get the latest handled tid from object_inrefs or 0 if table is empty
    """
    stmt = """
    SELECT DISTINCT tid FROM object_inrefs ORDER BY tid DESC LIMIT 1;
    """
    cursor.execute(stmt)
    if not cursor.rowcount:
        return 0
    (tid,) = cursor.fetchone()
    log.debug("boundary transaction id to start with is: tid=%d" % tid)
    return tid


def next_tid(cursor, lasttid):
    """get next higher tid after given lasttid
    """
    stmt = """
    SELECT DISTINCT tid
    FROM object_state
    WHERE tid > %d
    ORDER BY tid ASC
    LIMIT 1
    """ % lasttid
    cursor.execute(stmt)
    if not cursor.rowcount:
        return None
    (tid,) = cursor.fetchone()
    log.debug("next transaction id to process is: tid=%d" % tid)
    return tid

def changed_tids_len(cursor, tid):
    stmt = "SELECT COUNT(distinct tid) FROM object_state WHERE tid >=%d;" % tid
    cursor.execute(stmt)
    (count,) = cursor.next()
    return count or 0


################################################################################
# Creation/ update of inverse references table and counters

def _add_ref(cursor, source_zoid, target_zoid, tid):
    """insert an entry to the refcount table or update tid on existing

    - on source_zoid there is reference to target_zoid
    - so we have source_zoid as incoming reference on target_zoid
    """
    stmt = """
    SELECT add_inref(%(source_zoid)d, %(target_zoid)d, %(tid)d);
    """ % {'source_zoid': source_zoid, 'target_zoid': target_zoid, 'tid': tid}
    cursor.execute(stmt)


def _check_removed_refs(cursor, source_zoid, target_zoids):
    """get all prior filed references of current source_zoid
       and remove any not valid anymore, in other words if there is an entry in
       ``object_inrefs`` with inref=source_zoid but its zoid is not in
       target_zoids, remove it and decrement the counter for zoid.
    """
    stmt = """
    SELECT zoid
    FROM object_inrefs
    WHERE inref = %(source_zoid)s
    AND zoid <> %(source_zoid)s;
    """ % {'source_zoid': source_zoid}
    cursor.execute(stmt)
    stmt = ""
    for zoid in cursor:
        zoid = zoid[0]
        if zoid in target_zoids:
            continue
        log.debug(
            '    -> remove zoid=%d, inref=%d from object_inrefs' %
            (zoid, source_zoid)
        )
        stmt += """
        DELETE FROM object_inrefs
        WHERE zoid = %(zoid)s
        AND inref = %(source_zoid)s;

        UPDATE object_inrefs
        SET numinrefs = numinrefs - 1
        WHERE zoid = %(zoid)s
        AND inref = %(zoid)s;
        """ % {'source_zoid': source_zoid,
               'zoid': zoid}
    if stmt:
        cursor.execute(stmt)

@dbcommit
def handle_transaction(cursor, tid, initialize):
    """analyze a given transaction and fill inverse references
    """
    log.debug('handle transaction %d' % tid)
#    SET TRANSACTION ISOLATION LEVEL SERIALIZABLE;
    stmt = """
    BEGIN;
    SELECT zoid, state
    FROM object_state
    WHERE tid = %d
    ORDER BY zoid;
    """ % tid
    cursor.execute(stmt)
    zoid_count = 0
    refs_count = 0
    # cursor is needed, so store in array
    result = [(zoid, get_references(state)) for zoid, state in cursor]
    for source_zoid, target_zoids in result:
        log.debug('-> processing zoid=%d' % (source_zoid))
        log.debug('   found %d refs' % len(target_zoids))
        zoid_count += 1
        _add_ref(cursor, source_zoid, source_zoid, tid)
        for target_zoid in target_zoids:
            refs_count += 1
            log.debug('   -> process reference to %s' % target_zoid)
            _add_ref(cursor, target_zoid, target_zoid, tid)
            _add_ref(cursor, source_zoid, target_zoid, tid)

        if not initialize:
            _check_removed_refs(cursor, source_zoid, target_zoids)

    # commit whole handled tid, so we are sure to have it complete in numinrefs
    # funny part: this is multiple times faster than commit per source_zoid!
    return {'numzoids': zoid_count, 'numrefs': refs_count}

################################################################################
# Removal of orphaned objects

def _get_orphaned_zoid(cursor):
    stmt = """
    SELECT zoid
    FROM object_inrefs
    WHERE numinrefs = 1
    LIMIT 1;
    """
    cursor.execute(stmt)
    if not cursor.rowcount:
        return None
    (zoid,) = cursor.fetchone()
    log.debug("selected orphaned object: zoid=%d" % zoid)
    return zoid


def _remove_blob(storage, zoid):
    if storage.blobhelper is None:
        log.debug('-> No blobstorage available')
        return
    fshelper = storage.blobhelper.fshelper
    blobpath = fshelper.getPathForOID(p64(zoid))
    log.debug('-> Blobs for zoid=%s are at %s' % (zoid, blobpath))
    if not os.path.exists(blobpath):
        log.debug('-> No Blobs to remove')
        return
    log.debug('-> Remove Blobs')
    shutil.rmtree(blobpath)
    # here traversal up and removal of empty directories makes sense.
    # otherwise there are many empty directories left.
    # need to check side effects first!


def _remove_zoid(cursor, zoid):
    """
    remove a zoid completly.
    - remove all references in object_inrefs (this includes self reference)
    - remove entry in object_state
    """
    # handle references (remove, decrement)
    stmt = """
    SELECT zoid, state
    FROM object_state
    WHERE zoid = %d;
    """ % zoid
    cursor.execute(stmt)

    source_zoid, state = cursor.fetchone()
    target_zoids = get_references(state)
    stmt = ""
    for target_zoid in target_zoids:
        stmt += """
        DELETE FROM object_inrefs
        WHERE zoid = %(target_zoid)s
        AND inref = %(source_zoid)s;

        UPDATE object_inrefs
        SET numinrefs = numinrefs - 1
        WHERE zoid = %(target_zoid)s
        AND inref = %(target_zoid)s;
        """ % {'source_zoid': source_zoid,
               'target_zoid': target_zoid}

    # finally delete data
    stmt += """
    DELETE FROM object_inrefs
    WHERE zoid =  %(zoid)s;

    DELETE FROM object_state
    WHERE zoid =  %(zoid)s;
    """ % {'zoid': zoid}
    cursor.execute(stmt)


def remove_orphans(connection, cursor, storage):
    """remove orphans with blobs

    do transactions in here manually, because of blobs
    """
    tick = time.time()
    count = 0
    while True:
        zoid = _get_orphaned_zoid(cursor)
        if zoid is None:
            break
        log.debug('-> Remove orphaned with zoid=%s' % zoid)
        try:
            _remove_zoid(cursor, zoid)
            cursor.close()
            connection.commit()
        except:
            connection.rollback()
            raise
        _remove_blob(storage, zoid)
        count += 1

        if (count % CYCLES_TO_RECONNECT) == 0:
            # get a fresh connection, else postgres server may consume too
            # much RAM .oO( sigh )
            log.info('Refresh connection after {0} zoid cycles'.format(count))
            connection.close()
            connection, cursor = get_conn_and_cursor(storage)
        else:
            # only refresh closed cursor
            cursor = connection.cursor()


        cursor = connection.cursor()
        if (time.time() - tick) > 5:
            log.info('Removed %s orphaned objects' % count)
            tick = time.time()
    log.info('finished removal of %s orphaned objects' % count)
    return count

################################################################################
# Statistics

LOGLINE_TPL = """\
{tid_ratio:3.3f}% in {duration} | eta {eta} in {etadelta} | \
{processed_tids:d} of {overall_tids:d} ({tid_todo:d} todo) | \
{tid_rate:5.1f} t/s | \
{tid_delta_period:3d} t/delta ({tid_rate_period:5.1f} t/s) | \
{processed_zoids:d} zoids | {processed_refs} refs\
"""

def process_statistics(stats, force=False):
    if not force \
       and time.time() - stats['logtime'] < LOG_INTERVAL_SECS:
        return
    now = time.time()
    period = now - stats['logtime']

    duration = datetime.timedelta(seconds=now - stats['start'])
    stats['duration'] = str(duration).rsplit('.', 1)[0]

    stats['tid_delta_period'] = \
        stats['processed_tids'] - stats['processed_tids_offset']

    stats['tid_rate_period'] = stats['tid_delta_period'] / period

    stats['tid_ratio'] = \
        stats['processed_tids'] / float(stats['overall_tids']) * 100

    stats['tid_todo'] = stats['overall_tids'] - stats['processed_tids']
    stats['tid_rate'] = stats['processed_tids'] / (now - stats['start'])

    stats['left'] = \
        ((now - stats['start']) / stats['processed_tids']) * stats['tid_todo']

    stats['left'] = datetime.timedelta(seconds=stats['left'])

    stats['eta'] = \
        (datetime.datetime.now() + stats['left']).strftime('%Y-%m-%d %H:%M')
    stats['etadelta'] = str(stats['left']).rsplit('.', 1)[0]

    log.info(LOGLINE_TPL.format(**stats))

    stats['logtime'] = now
    stats['processed_tids_offset'] = stats['processed_tids']


################################################################################
# Main Runner

def run(argv=sys.argv):
    parser = optparse.OptionParser(
        description='Fast ZODB Relstorage Packer for history free PostgreSQL',
        usage="%prog config_file"
    )
    parser.add_option(
        "-i", "--init", dest="initialize", default=False,
        action="store_true",
        help="Removes all reference counts and starts from scratch.",
    )
    parser.add_option(
        "-v", "--verbose", dest="verbose", default=False,
        action="store_true",
        help="More verbose output, includes debug messages.",
    )
    options, args = parser.parse_args(argv[1:])
    if len(args) != 1:
        parser.error("The name of one configuration file is required.")
    if options.verbose:
        log.setLevel(logging.DEBUG)
        log.debug("Logging in verbose mode.")

    log.info("Initiating packing.")

    storage = get_storage(args[0])
    connection, cursor = get_conn_and_cursor(storage)
    aquire_lock(connection, cursor)
    cursor = connection.cursor()

    if options.initialize:
        try:
            init_table(connection, cursor)
            cursor = connection.cursor()
        except:
            release_lock(connection, cursor)
            storage.close()
            raise
    stats = {
        'processed_tids': 0,
        'processed_zoids': 0,
        'processed_refs': 0,
    }
    stats['start'] = stats['logtime'] = time.time()
    try:
        initialize = options.initialize
        init_tid = tid = tid_boundary(cursor)
        stats['processed_tids_offset'] = 0
        if initialize:
            log.info('Fetching number of all transactions from DB ...')
        else:
            log.info(
                "Fetching number of new transactions since tid {0} "
                "from DB ...".format(init_tid)
            )
        stats['overall_tids'] = changed_tids_len(cursor, init_tid)
        if initialize:
            log.info('-> {overall_tids} transactions in DB'.format(**stats))
        else:
            log.info('-> {overall_tids} new transactions in DB'.format(**stats))

        # BUILD/ UPDATE INVERSE REFERENCES
        while True:
            tid = next_tid(cursor, tid)
            if not tid:
                break

            # BUILD/UPDATE FOR TID
            handle_stats = handle_transaction(
                connection,
                cursor,
                tid,
                initialize=initialize
            )
            stats['processed_tids'] += 1
            stats['processed_zoids'] += handle_stats['numzoids']
            stats['processed_refs'] += handle_stats['numrefs']
            if (stats['processed_tids'] % CYCLES_TO_RECONNECT) == 0:
                # get a fresh connection, else postgres server may consume too
                # much RAM .oO( sigh )
                connection.close()
                log.info(
                    'Refresh connection after {processed_tids} tid '
                    'cycles'.format(
                        **stats
                    )
                )
                connection, cursor = get_conn_and_cursor(storage)
            else:
                # only refresh closed cursor
                cursor = connection.cursor()

            # Statistics
            process_statistics(stats)
        if stats['processed_tids']:
            process_statistics(stats, True)
        processing_time = time.time() - stats['start']
        log.info(
            'Finished analyzation phase after %s (%.2fs)' %
            (str(datetime.timedelta(seconds=processing_time)), processing_time)
        )
        cleanup_start = time.time()

        # REMOVE
        removed_count = remove_orphans(connection, cursor, storage)

        processing_time = time.time() - cleanup_start
        log.info(
            'Finished cleanup phase after %s (%.2fs)' %
            (str(datetime.timedelta(seconds=processing_time)), processing_time)
        )
    except Exception, e:
        log.error(e.message)
        raise
        exit(1)
    finally:
        # check if connection is closed!
        try:
            cursor = connection.cursor()
        except InterfaceError:
            connection, cursor = get_conn_and_cursor(storage)
        release_lock(connection, cursor)
        connection.close()
        storage.close()

    if stats['processed_tids']:
        processing_time = time.time() - stats['start']
        log.info(
            "Completed in {mode}-mode: processed {processed_tids} tids, "
            "{processed_zoids} zoids, {processed_refs} refs, "
            "removed {remove_count} objects, "
            "took {processing_time} ({processing_time_secs:.2f}s) ".format(
                mode='init' if options.initialize else 'update',
                remove_count=removed_count,
                processing_time_secs=processing_time,
                processing_time=str(
                    datetime.timedelta(seconds=processing_time)
                ),
                **stats
            )
        )
    else:
        log.info("Completed, there was nothing to do.")
